### PR TITLE
Fixed #9762 -- Made DateFormat.r() locale-independent.

### PR DESCRIPTION
**Pull request by bmispelon**
Thanks to Antonio Melé for the original report all those years ago
and to all the contributors who helped along the way.

ticket-9762

### CODE REVIEW

File: django/utils/dateformat.py [hunk 1]
```diff
@@ -202,6 +205,11 @@ def Z(self):
         # Positive offsets have days=0
         return offset.days * 86400 + offset.seconds
 
+    def r(self):
+        # Implement a stub here so that using the `r` format with a date object
+        # raises the correct error.
+        raise NotImplementedError
```
Comment (felixxm) [inline django/utils/dateformat.py]: This is not strictly related with a bugfix. Can we move this to a separate commit? also test is necessary.

File: django/utils/dateformat.py [hunk 2]
```diff
@@ -202,6 +205,11 @@ def Z(self):
         # Positive offsets have days=0
         return offset.days * 86400 + offset.seconds
 
+    def r(self):
+        # Implement a stub here so that using the `r` format with a date object
+        # raises the correct error.
```
Comment (felixxm) [inline django/utils/dateformat.py]: I'm not sure if this comment is necessary.

File: tests/utils_tests/test_dateformat.py [hunk 3]
```diff
@@ -149,10 +149,18 @@ def test_timezones(self):
...
             msg = (
                 "The format for date objects may not contain time-related "
                 "format specifiers (found '%s')." % specifier
             )
             with self.assertRaisesMessage(TypeError, msg):
                 dateformat.format(my_birthday, specifier)
+
+    def test_r_format_with_non_en_locale(self):
+        """
+        #9762 -- Changing the locale should not change the `r` format.
```
Comment (felixxm) [inline tests/utils_tests/test_dateformat.py]: We don't refer tickets in tests anymore, please chop `#9762`, also maybe:
```
# Changing the locale doesn't change the "r" format.
```

File: django/utils/dateformat.py [hunk 4]
```diff
@@ -278,7 +286,15 @@ def o(self):
 
     def r(self):
         "RFC 5322 formatted date; e.g. 'Thu, 21 Dec 2000 16:01:07 +0200'"
-        return self.format('D, j M Y H:i:s O')
+        if not self.timezone:
+            raise TypeError("The 'r' format specifier can only be used on datetime objects")
```
Comment (felixxm) [inline django/utils/dateformat.py]: Do we need this? I don't see any failures without it, I think we can simplify this method, e.g.
```python
def r(self):
    if self.timezone and is_naive(self.data):
         self.data = make_aware(self.data, timezone=self.timezone)
    return format_datetime_rfc5322(self.data)
```

File: django/utils/dateformat.py (hunk 1 above)
Comment (bmispelon) [reply django/utils/dateformat.py]: I've moved this to a separate commit. The test was done simply by adding `'r'` to the list of formats tested in `test_invalid_time_format_specifiers`. That should be enough, no?

File: django/utils/dateformat.py (hunk 4 above)
Comment (bmispelon) [reply django/utils/dateformat.py]: The `raise TypeError(...)` was leftover from a previous approach I'd tried, I've removed it.

Your suggested approach of mutating `self.data` could lead to problems because the `self.data` object is reused for all the format characters passed to `.format()`.

Consider the following example: `format(datetime.now(), 'U r U')`.
The first `U` format will receive a naive datetime object, then the `r` format will mutate that value to be an aware one, then finally the second `U` format will receive the new aware object.

I'm not sure if it can cause actual problems in practice but it's giving me a bad feeling (and none of the other format methods manipulate `self.data`).

Comment (felixxm) [reply django/utils/dateformat.py]: Agreed :+1:

File: django/utils/dateformat.py (hunk 2 above)
Comment (bmispelon) [reply django/utils/dateformat.py]: I figured it might be a bit confusing to see a `NotImplemented` there but only for `r()`.

As far as I can tell, `r()` is the only format that strictly accepts datetime objects and not a combination of datetime/date or datetime/time.

File: docs/releases/3.1.txt [hunk 5]
```diff
@@ -228,7 +228,9 @@ Signals
 Templates
 ~~~~~~~~~
 
-* ...
+* The ``r`` format for the :tfilter:`date` filter (and the :ttag:`now` tag)
+  is unaffected by the current locale and will always use English (as per
+  :rfc:`5322`).
```
Comment (felixxm) [inline docs/releases/3.1.txt]: Release notes are not necessary, because it's a bug not a new feature.

File: django/utils/dateformat.py (hunk 2 above)
Comment (felixxm) [reply django/utils/dateformat.py]: Agreed, it can be confusing.

Comment (felixxm) [reply django/utils/dateformat.py]: TBH even now it's confusing :thinking: _"explicit is better than implicit"_ 🧘‍♂️ so I decided to move it to the `DateFormat.r()`.

**Comment by felixxm** [comment]
@bmispelon Thanks again :+1: I pushed minor edits.

**Comment by bmispelon** [comment]
Excellent, thanks :+1: `

Is there a particular reason why you went with `if type(self.data) is datetime.date` instead of `if isinstance(self.data, datetime.date)`?

**Comment by felixxm** [comment]
Yes, because `datetime.datetime()` is an instance of `datetime.date`.

**Comment by bmispelon** [comment]
Oh, I hadn't thought of that :man_facepalming:
